cityofaustin/atd-moped
# Upgrade draw tools to `mapbox-gl-draw` to support `react-map-gl` v7

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/9707

`react-map-gl-draw` -> `@mapbox/mapbox-gl-draw` plugin

## Testing
**URL to test:** 
https://md-9707-mapbox-gl-draw--atd-moped-main.netlify.app/moped/

**Steps to test:**
1. Draw and save project components features:
   - Point (bike box)
   - Line (bike lane)
2. After saving, click a component's arrow icon and the interface will load existing drawn features to be removed or edited. You are now ready for step 3.
<img width="334" alt="Screen Shot 2022-08-22 at 11 15 27 AM" src="https://user-images.githubusercontent.com/37249039/185969338-acc7cbd6-e6bd-40a9-93b8-63cbee7cfa7a.png">

3. To delete a drawn feature and update with a new feature, click a drawn feature and you should see the point outline get wider or the line get wider with the vertices highlighted with green circles. Once highlighted, click the trash icon to delete the feature. Draw a replacement feature and click "Save", and you will see the old feature removed and the new feature saved to the project.
<img width="169" alt="Screen Shot 2022-08-22 at 11 32 40 AM" src="https://user-images.githubusercontent.com/37249039/185973202-c346692e-3acb-47c2-ae59-130b1b08760f.png">
<img width="240" alt="Screen Shot 2022-08-22 at 11 37 15 AM" src="https://user-images.githubusercontent.com/37249039/185973579-3f8160fe-88c9-45f6-8080-04e8baf843a1.png">

4. You can cancel a create, edit, or delete by clicking the "Cancel" button in the components toolbar
<img width="528" alt="Screen Shot 2022-08-22 at 11 22 47 AM" src="https://user-images.githubusercontent.com/37249039/185970633-50b03421-9bc0-4812-8910-0ca1f383b709.png">

5. You should see a warning that at least one feature must be chosen when:
   - a feature has not been drawn yet
   - a single drawn feature is deleted without drawing a replacement or selecting a replacement from a selectable layer
<img width="328" alt="Screen Shot 2022-08-22 at 11 27 41 AM" src="https://user-images.githubusercontent.com/37249039/185971563-c8488856-d6dd-475b-9ea1-4d78b66b3618.png">

6. Active control button background is blue when drawing
<img width="128" alt="Screen Shot 2022-08-22 at 11 31 33 AM" src="https://user-images.githubusercontent.com/37249039/185972278-e91f4d18-9846-46a0-bb58-670531b023f2.png">

7. You can draw multiple points/lines in one edit of an existing component. The draw mode should exit after each new feature that you draw. You can also delete multiple points/lines in one edit of an existing component.
8. You can toggle drawn points, drawn lines, lines, and points visibility with the "Map Features" layer dropdown when not in edit mode.
9. While drawing (cursor shows as a crosshair), the selectable street and point layers are not interactive
10. When not drawing, the selectable street and point layers are interactive

## Component feature formats

### Question: should we stick with [uuidv4](https://github.com/thenativeweb/uuidv4) for avoiding drawn feature ID collisions or switch over to [what Mapbox GL Draw does out of the box](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/feature_types/feature.js#L8) with [hat](https://github.com/substack/node-hat)? More info below.

### `react-map-gl-draw` - what we use now
unique IDs created by [uuidv4](https://github.com/thenativeweb/uuidv4)
```md
{
"id": "fae02dfb-02b4-49b1-8758-ab72b906182a",
"type": "Feature", 
"geometry": {
      "type": "LineString", 
      "coordinates": [ [-97.74843629360308, 30.26634733277872], [-97.74800714289084, 30.26599548905569] ]
      }, 
"properties": {
      "renderType": "LineString", 
      "sourceLayer": "drawnByUserLine", 
      "PROJECT_EXTENT_ID": "fae02dfb-02b4-49b1-8758-ab72b906182a"
      }
}
```

### `mapbox-gl-draw` - what this PR introduces
unique IDs created by [hat](https://github.com/substack/node-hat)
```md
{ 
"id": "af25be770fd2843dba8dba80306c2927",
"type": "Feature", 
"geometry": { 
      "type": "LineString", 
      "coordinates": [ [ -97.6854132744357, 30.289124655635447 ], [ -97.68675452987974, 30.287097850485324 ] ] 
      }, 
"properties": { 
      "renderType": "LineString", 
      "sourceLayer": "drawnByUserLine", 
      "PROJECT_EXTENT_ID": "af25be770fd2843dba8dba80306c2927"
      } 
}
```

---
#### Ship list
